### PR TITLE
RavenDB-15453

### PR DESCRIPTION
**Pull request by ayende**
- Fixing OOM when during recursive graph query on large graph
- Fixing tx already open when computing etag on query after the data was changed.

### CODE REVIEW

File: test/SlowTests/Graph/RavenDB-15453.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace SlowTests.Graph
+{
+    public class RavenDB_15453 : RavenTestBase
+    {
+        public RavenDB_15453(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        public class Item
```
Comment (ppekrol) [inline test/SlowTests/Graph/RavenDB-15453.cs]: private

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 2]
```diff
@@ -3448,9 +3448,17 @@ public long GetIndexEtag(QueryOperationContext context, QueryMetadata q)
                 using (_contextPool.AllocateOperationContext(out TransactionOperationContext indexContext))
                 {
                     using (indexContext.OpenReadTransaction())
-                    using (context.OpenReadTransaction())
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Index.cs:3451]: I don ot like this change. You should use 'CalculateIndexEtag' directly. IIRC the order of opened transactions heree mattered.

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 3]
```diff
@@ -3448,9 +3448,17 @@ public long GetIndexEtag(QueryOperationContext context, QueryMetadata q)
...
                     {
-                        return CalculateIndexEtag(context, indexContext, q, IsStale(context, indexContext));
+                        DocumentsTransaction  d = null;
+                        if (context.Documents.HasTransaction == false)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Index.cs]: This is incorrect. QueryOperationContext operates on both Server-Storage and Database-Storage transactions. IIRC there is a method there 'AreTransactionsOpened' to check that.

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 4]
```diff
@@ -3448,9 +3448,17 @@ public long GetIndexEtag(QueryOperationContext context, QueryMetadata q)
                 using (_contextPool.AllocateOperationContext(out TransactionOperationContext indexContext))
                 {
                     using (indexContext.OpenReadTransaction())
-                    using (context.OpenReadTransaction())
                     {
-                        return CalculateIndexEtag(context, indexContext, q, IsStale(context, indexContext));
+                        DocumentsTransaction  d = null;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Index.cs]: `d` ?

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 3 above)
Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Index.cs]: Changed to use that

File: test/SlowTests/Graph/RavenDB-15453.cs (hunk 1 above)
Comment (ayende) [reply test/SlowTests/Graph/RavenDB-15453.cs]: done

**Review by ppekrol — APPROVED** [review]

**Comment by ppekrol** [comment]
It causes the 'BasicGraphQueries.CanHandleCyclesInGraph' to fail, altough the asssertion there looks suspicious? https://jenkins.hibernatingrhinos.com/job/PR_Tests_v5.0/1430/testReport/junit/(root)/(empty)/Graph_BasicGraphQueries_CanHandleCyclesInGraph/

```
Assert.Equal() Failure\r\nExpected: String[] [\"employees/5-A\", \"employees/2-A\", \"employees/2-A\"]\r\nActual:   String[] [\"employees/5-A\", \"employees/2-A\"]
```

**Comment by ayende** [comment]
Yes, the test itself is wrong here.

**Pull request merged**